File: python/leetcode034.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ls = len(nums)
        if ls == 1 and nums[0]==target:
            return [0,0]
        i = 0
        k = ls-1
        while i <= k:
            mid = (i + k) // 2
            if nums[mid] == target:
                l = r = mid
                while l > 0 and nums[l-1] == nums[l]:
                    l = l - 1
                while r < ls-1 and nums[r+1] == nums[r]:
                    r = r + 1
                return [l,r]
            elif nums[mid] > target:
                k = mid-1
            else:
                i = mid+1
        return [-1,-1]

if __name__ == "__main__":
    s = Solution()
    print(s.searchRange([1,1,2],1))

File: leetcodec++/leetcode034.cpp
//
// Created by 黄敏辉 on 2019-07-10.
//

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int start = 0, end = nums.size() - 1;
        int mid;
        while(start <= end){
            mid = (start + end) / 2;
            if(nums[mid] == target){
                int l = mid, r = mid;
                while(nums[start] != target && start < l){
                    if (nums[(start + l) / 2] == target)
                        l = (start + l) / 2;
                    else
                        start = (start + l) / 2 + 1;
                }
                while(nums[end] != target && r < end){
                    if (nums[(r + end + 1) / 2] == target)
                        r = (r + end + 1) / 2;
                    else
                        end = (r + end + 1) / 2 - 1;
                }
                return vector<int> {start, end};
            }
            else if(nums[mid] < target){
                start = mid + 1;
            }
            else{
                end = mid - 1;
            }
        }
        return vector<int> {-1, -1};
    }
};